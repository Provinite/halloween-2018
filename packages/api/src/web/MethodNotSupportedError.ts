/**
 * Error indicating that the requested method is not supported for this route.
 */
export class MethodNotSupportedError extends Error {}
