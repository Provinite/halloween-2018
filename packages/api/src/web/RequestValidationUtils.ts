import * as moment from "moment";

export type ValidationResult = {
  pass: boolean;
  message: string;
};
export type ValidatorFunction<Keys extends string> = (
  val: any,
  key: string,
  obj?: Partial<Record<Keys, string>>
) => ValidationResult;
export type TypedValidatorFunction<Keys extends string, R> = ValidatorFunction<
  Keys
> & {
  __returntype: R;
};
export type ValidationFor<Keys extends string> = {
  [key in Keys]:
    | ValidatorFunction<Keys>
    | undefined
    | TypedValidatorFunction<Keys, any>
};
export type ValidatedRequestBody<V extends ValidationFor<any>> = {
  [key in keyof V]: V[key] extends TypedValidatorFunction<any, infer U>
    ? U
    : any
};
/**
 * Validate an incoming request body.
 * @param requestBody - The incoming request body.
 * @param validators - A map of validator functions for each key that will
 *  be picked off of requestBody.
 * @throws RequestValidationError if any validators fail.
 */
export function validateRequest<
  Keys extends string,
  T extends ValidationFor<Keys>
>(requestBody: any, validators: T): ValidatedRequestBody<T> {
  const alwaysPass = () => validationResult(true, "No validation rules.");
  const keys = Object.keys(validators) as Keys[];
  const body = pickRequestKeys(requestBody, ...keys);
  const errorResults: Partial<Record<Keys, ValidationResult>> = {};
  let failed = false;
  for (const key of keys) {
    const validator =
      validators[key] === undefined ? alwaysPass : validators[key];
    const result = validator(requestBody[key], key, body);
    if (result.pass !== true) {
      failed = true;
      errorResults[key] = result;
    }
  }
  if (failed === true) {
    throw new RequestValidationError(errorResults);
  }
  return body as any;
}

export function validateValue(
  value: any,
  name: string,
  validator: (val: any, name: string) => ValidationResult
) {
  const result = validator(value, name as any);
  if (result.pass === false) {
    throw new RequestValidationError({ [name]: result });
  }
  return value;
}

/**
 * Used to select specific known keys from an incoming request body.
 * @param requestBody - Cannot be null or undefined.
 * @param keys - The keys to extract.
 * @return An object containing the specified keys from requestBody.
 */
export function pickRequestKeys<T, K extends string>(
  requestBody: any,
  ...keys: K[]
): Partial<Record<K, any>> {
  const result = {} as any;
  for (const key of keys) {
    if (requestBody.hasOwnProperty(key)) {
      result[key] = requestBody[key];
    }
  }
  return result;
}

/**
 * Create a validation result.
 * @param pass - True if the test passed.
 * @param message - The message to display for failures.
 */
export function validationResult(
  pass: boolean,
  message: string
): ValidationResult {
  return {
    pass,
    message
  };
}

export type CreateValidatorOptions<T> = {
  optional: boolean;
  nullable: boolean;
  validator: TypedValidatorFunction<string, T>;
};
export function createValidator<R = any>(
  options: CreateValidatorOptions<R>
): TypedValidatorFunction<string, R> {
  return ((val: any, key: string, obj: any) => {
    if (val === undefined) {
      return validationResult(
        options.optional,
        `Missing required field: "${key}".`
      );
    }
    if (val === null) {
      return validationResult(options.nullable, `"${key}" cannot be empty.`);
    }
    return options.validator(val, key, obj);
  }) as any;
}

/**
 * Determine if the value is a string of only or more digits.
 * @param val - The value to test.
 * @return true iff val is a string that consists only of one or more digits.
 */
export function isDigitString(val: any) {
  if (typeof val !== "string") {
    return false;
  }
  return /^\d+$/.test(val);
}

/**
 * Check if a string consists
 * @param val
 */
export function isValidInt(val: any) {
  if (typeof val !== "number") {
    return false;
  }
  return Number.isInteger(val);
}

export function isValidFloat(val: any) {
  return typeof val === "number" && !Number.isNaN(val);
}

export function isPresent<T>(val: T): val is Exclude<T, undefined> {
  return val !== undefined;
}

export function isString(val: any): val is string {
  return typeof val === "string";
}

export function isNonEmptyString(val: any): boolean {
  return typeof val === "string" && val.length > 0;
}

export function isValidDateString(val: any): val is string {
  if (typeof val !== "string" || !val) {
    return false;
  }
  return moment(val, "YYYY-MM-DD", true).isValid();
}

export function isValidDateTimeString(val: any): val is string {
  if (typeof val !== "string") {
    return false;
  }
  return moment(val, "YYYY-MM-DD HH:mm:ss", true).isValid();
}

/**
 * Error thrown to indicate a request validation failed.
 */
export class RequestValidationError<Keys extends string> extends Error {
  constructor(
    /**
     * Each key that failed validation will have a ValidationResult in this
     * object.
     */
    public erroredResults: { [key in Keys]?: ValidationResult }
  ) {
    super(
      `Validation failed for field(s): ${Object.keys(erroredResults).join(
        ", "
      )}`
    );
  }
}

const requiredString = (val: any, key: string) => {
  return validationResult(
    isString(val),
    `${key} is required and must be a string.`
  );
};
const requiredNonEmptyString = (val: any, key: string) =>
  validationResult(isNonEmptyString(val), `${key} must be a non empty string`);
const requiredDateString = (val: any, key: string) => {
  const result = isValidDateString(val);
  return validationResult(result, `${key} must be a valid date (YYYY-MM-DD).`);
};

const requiredFloat = (val: any, key: string) => {
  const result = isValidFloat(val);
  return validationResult(result, `"${key}" must be a valid decimal number.`);
};
const requiredDigitString = (val: any, key: string) => {
  return validationResult(
    isDigitString(val),
    `${key} must be a string of only digits.`
  );
};
const requiredInt = (val: any, key: string) => {
  return validationResult(
    isValidInt(val),
    `"${key}" must be a valid whole number.`
  );
};
const requiredUndefined = (val: any, key: string) => {
  return validationResult(val === undefined, `Field not allowed: "${key}"`);
};
const rootValidators = {
  integer: requiredInt as TypedValidatorFunction<string, number>,
  digitString: requiredDigitString as TypedValidatorFunction<string, string>,
  string: requiredString as TypedValidatorFunction<string, string>,
  nonEmptyString: requiredNonEmptyString as TypedValidatorFunction<
    string,
    string
  >,
  notAllowed: requiredUndefined as TypedValidatorFunction<string, never>,
  float: requiredFloat as TypedValidatorFunction<string, number>,
  dateString: requiredDateString as TypedValidatorFunction<string, string>
};

type RootValidators = typeof rootValidators;
type OptionalValidators = {
  [key in keyof RootValidators]: RootValidators[key] extends TypedValidatorFunction<
    string,
    infer U
  >
    ? TypedValidatorFunction<string, U | undefined>
    : never
} & {
  nullable: OptionalAndNullableValidators;
};
type NullableValidators = {
  [key in keyof RootValidators]: RootValidators[key] extends TypedValidatorFunction<
    string,
    infer U
  >
    ? TypedValidatorFunction<string, U | null>
    : never
} & {
  optional: OptionalAndNullableValidators;
};
type OptionalAndNullableValidators = {
  [key in keyof RootValidators]: RootValidators[key] extends TypedValidatorFunction<
    string,
    infer U
  >
    ? TypedValidatorFunction<string, U | null | undefined>
    : never
};
type OptionalNullableValidators = RootValidators & {
  optional: OptionalValidators;
  nullable: NullableValidators;
};

const optionalValidators = {} as any;
const nullableValidators = {} as any;
const optionalNullableValidators = {} as any;
for (const [key, validator] of Object.entries(rootValidators)) {
  optionalValidators[key] = createValidator({
    optional: true,
    nullable: false,
    validator: validator as any
  });
  nullableValidators[key] = createValidator({
    optional: false,
    nullable: true,
    validator: validator as any
  });
  optionalNullableValidators[key] = createValidator({
    optional: true,
    nullable: true,
    validator: validator as any
  });
}
export const validators: OptionalNullableValidators = {
  ...rootValidators,
  optional: {
    ...optionalValidators,
    nullable: optionalNullableValidators
  },
  nullable: {
    ...nullableValidators,
    optional: optionalNullableValidators
  }
};
