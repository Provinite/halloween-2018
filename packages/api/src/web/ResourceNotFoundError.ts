/**
 * Error thrown indicating a requested resource was not found.
 */
export class ResourceNotFoundError extends Error {}
