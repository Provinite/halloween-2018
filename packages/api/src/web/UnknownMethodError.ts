/**
 * Error indicating the request method is unimplemented.
 */
export class UnknownMethodError extends Error {}
