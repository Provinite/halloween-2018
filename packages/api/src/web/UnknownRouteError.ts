/**
 * Error thrown indicating that the requested route is not known to the
 * application.
 */
export class UnknownRouteError extends Error {}
