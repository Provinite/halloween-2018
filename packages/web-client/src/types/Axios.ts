import Axios from "axios";
import { ArgumentTypes } from "./ArgumentTypes";

export type InterceptorDefinition = ArgumentTypes<
  typeof Axios.interceptors.response.use
>;
export const noOnFulFilled = <T>(result: T): T => result;
export const noOnError = (err: any): never => {
  throw err;
};
