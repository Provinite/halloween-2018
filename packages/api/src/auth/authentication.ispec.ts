import Axios, { AxiosError, AxiosInstance } from "axios";
import { RequestValidationError } from "../web/RequestValidationUtils";

describe("/login", () => {
  let axios: AxiosInstance;
  beforeEach(() => {
    axios = Axios.create();
    axios.defaults.baseURL = "http://localhost:8081/";
  });
  describe("POST /login", () => {
    describe("error handling", () => {
      const invalidAuthCodes = [undefined, 123, null];
      it.each([invalidAuthCodes])(
        "returns a 400/RequestValidationError for bad auth codes",
        async authCode => {
          let caughtError: AxiosError;
          await axios
            .post("/login", { authCode })
            .catch(e => (caughtError = e));
          expect(caughtError).toBeTruthy();
          expect(caughtError.response.status).toBe(400);
          expect(caughtError.response.data.error).toEqual(
            RequestValidationError.name
          );
          expect(caughtError.response.data).toMatchSnapshot();
        }
      );
    });
  });
});
