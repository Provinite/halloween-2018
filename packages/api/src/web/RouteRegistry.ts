import { Resolver } from "awilix";
import { RoleLiteral } from "../auth/RoleLiteral";
import { RouteTransformationService } from "../config/RouteTransformationService";
import { HttpMethod } from "../HttpMethod";
import { logger } from "../logging";
import { Component } from "../reflection/Component";
import { MethodNotSupportedError } from "./MethodNotSupportedError";
import { UnknownRouteError } from "./UnknownRouteError";
@Component()
/**
 * Class used for mapping request paths and http methods to handler resolvers.
 */
export class RouteRegistry {
  /** Data structure mapping route -> method -> resolver */
  private map: IRouteMap;
  private transformationService: RouteTransformationService;
  constructor(routeTransformationService: RouteTransformationService) {
    this.map = {};
    this.transformationService = routeTransformationService;
  }
  /**
   * Assign the supplied awilix resolver as the handler for the given
   * route and http verb(s).
   * @return this
   */
  registerRoute(
    route: string,
    methods: HttpMethod | HttpMethod[],
    resolver: Resolver<any> | ((...args: any[]) => any),
    router: any,
    allowedRoles: RoleLiteral[]
  ): this {
    logger.info(
      `RouteRegistry#registerRoute: Registering endpoint "${route}" [${methods}]`
    );
    if (!this.map[route]) {
      this.map[route] = {};
    }
    if (!Array.isArray(methods)) {
      methods = [methods];
    }
    methods.forEach(method => {
      this.map[route][method] = {
        router,
        resolver,
        allowedRoles
      };
    });
    return this;
  }

  /**
   * Get the resolver and path variables for the given request path.
   * @return The resolver and any path variables for the route and method.
   */
  lookupRoute(
    requestPath: string,
    method: HttpMethod
  ): {
    /**
     * The router instance for the handler, if any.
     */
    router: any;
    /**
     * An awilix resolver for the actual route handler. If lookup fails,
     * undefined.
     */
    resolver?: Resolver<any> | ((...args: any[]) => any);
    /**
     * A map of the extracted path variables for the request (name => value).
     * Undefined if lookup fails.
     */
    pathVariables?: { [key: string]: string };
    /**
     * A list of roles that may use this route.
     */
    allowedRoles?: RoleLiteral[];
  } {
    // Exact match (but don't get false exact matches on wildcard routes if the
    // request path literally contains something like {id})
    if (!requestPath.includes("{")) {
      if (this.map[requestPath]) {
        if (this.map[requestPath][method]) {
          return {
            router: this.map[requestPath][method].router,
            resolver: this.map[requestPath][method].resolver,
            allowedRoles: this.map[requestPath][method].allowedRoles
          };
        } else {
          throw new MethodNotSupportedError(Object.keys(
            this.map[requestPath]
          ) as HttpMethod[]);
        }
      }
    }

    // Check wildcard routes
    for (const route in this.map) {
      if (!route.includes("{")) {
        continue;
      }
      // TODO: This could be cached. Calculating on every request seems
      // wasteful.
      const parsedRoute = this.transformationService.parseRoute(route);
      const pathVariables = this.transformationService.getPathVariables(
        parsedRoute,
        requestPath
      );
      if (pathVariables) {
        // route matched, check method support
        if (this.map[route][method]) {
          return {
            router: this.map[route][method].router,
            resolver: this.map[route][method].resolver,
            allowedRoles: this.map[route][method].allowedRoles,
            pathVariables
          };
        } else {
          throw new MethodNotSupportedError(Object.keys(
            this.map[route]
          ) as HttpMethod[]);
        }
      }
    }
    // No matches found
    throw new UnknownRouteError();
  }
}

/* Private Interfaces */
/**
 * Object mapping http methods to resolvers for a given route.
 */
type IRouteHandler = {
  [method in HttpMethod]?: {
    router: any;
    resolver: Resolver<any> | ((...args: any[]) => any);
    allowedRoles: RoleLiteral[];
  }
};
/**
 * Object mapping routes to handler objects.
 */
interface IRouteMap {
  [route: string]: IRouteHandler | undefined;
}

declare global {
  interface ApplicationContext {
    /** The registry of all routes that this application covers */
    routeRegistry: RouteRegistry;
  }
}
