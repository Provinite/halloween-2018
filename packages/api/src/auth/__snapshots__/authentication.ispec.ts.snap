// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/login POST /login error handling returns a 400/RequestValidationError for bad auth codes 1`] = `
Object {
  "error": "RequestValidationError",
  "errors": Object {
    "authCode": Object {
      "message": "authCode is required and must be a string.",
      "pass": false,
    },
  },
  "message": "Validation failed for field(s): authCode",
  "status": 400,
}
`;
